### DISCUSSION TITLE
How to use netDXF - Basics

### DISCUSSION THREAD

**Discussion by atemmler** [haplokuon/netDxf#420] [Q&A]
Hello,
I think the netDXF Library looks very interesting for my application (laser processing research project). However, I am not only a beginner in using netDXF, but also in C# programming, so I could use some tips / examples on how to use the netDXF Library.

My task is comparatively simple:
1. load dxf file (which often consists of points and lines only)
2. count all elements (or all line elements)
3. write start and end points of the lines into an array

That means concretely my questions are: 
a) How do I search a loaded DXF file for specific elements (e.g. lines)?
b) How can I write the start and end points of ALL found line elements into an array?

Possibly the questions seem trivial, however I am still completely unclear how to work with the netDXF library.

Many thanks & greetings,
André.

**Comment by DavidWishengrad** [comment]
I do not have the time to help you atm thoroughly, but this is exactly what you are looking for and is super fast and Daniel does a great job.

No matter what you need to go through, W3 school, etc., get busy, have visual studio 2019 community edition installed (not sure about 21) and a working sample exe project made that can read and write a text file. 

Once you have that, you can then reference the netDXF runtime (not debug) dll, which is already compiled in the 4.5 release folder, and use that to do what you really want to do (knowing you have a vaild net application that you can compile and run.

Once you have your project referencing that dll, you will then import netDXF libraries (see end of this list):
using System;
using System.Runtime;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Runtime.InteropServices;
using System.Text;
using System.Windows.Forms;
using System.Text.RegularExpressions;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Linq;
**using netDxf;
using netDxf.Entities;
using netDxf.Blocks;
using netDxf.Collections;
using GTE = netDxf.GTE;
using netDxf.Header;
using netDxf.Objects;
using netDxf.Tables;
using netDxf.Units;
using Attribute = netDxf.Entities.Attribute;
using FontStyle = netDxf.Tables.FontStyle;
using Image = netDxf.Entities.Image;
using Point = netDxf.Entities.Point;
using Trace = netDxf.Entities.Trace;
using Vector2 = netDxf.Vector2;
using Vector3 = netDxf.Vector3;**

There are a great many code examples included with netDXF for opening and reading a file's content. Look and try hard and do your due diligence. Read everything in the code and everything written here. netDXF is very very good and well worth trying hard, no matter what. more later....after hearing you have those things installed and a simple project working without netDXF.

**Reply by atemmler** [reply]
Hey David, 

thanks a lot for your advise. I have a little bit of coding experience and already installed VS2022. I also implemented the netDXF.dll and my project/solution is using most of the netDXF libraries you suggested. Loading a dxf-file already works. I am actually struggling to find examples that help me to better understand how to read the dxf file's content. Particularly, scanning for specific objects and getting their start and end points.
If you could point me into the direction of one, or two great code examples, that would help a lot.

Thanks and best regards, 
André.

**Reply by DavidWishengrad** [reply]
Will any of this help you? There are probably error, as I was fixing some stuff there when I had to move on to other stuff and let it sit,
**Also, look at the program.cs in netDXF. That's the best resource I have found** other than asking Daniel directly. 


`DxfDocument loaded = DxfDocument.Load("C:\\Log\\fix.dxf");
                loaded.Entities.ActiveLayout = netDxf.Objects.Layout.ModelSpaceName;

                EntityCollection entities = loaded.Blocks[Block.DefaultModelSpaceName].Entities;

                int entitycount = -1;
                entitycount = entities.Count;

                int entitytimer = 0;
                entitytimer = entitycount / 10000;
                double vertsx = 0;
                double vertsy = 0;
                double vertsz = 0;
                double vertex = 0;
                double vertey = 0;
                double vertez = 0;
                int vertextcount = 0;
                double vertx1 = 0;
                double verty1 = 0;
                double vertx2 = 0;
                double verty2 = 0;
                int roundamount = 16;
                double totallength = 0;
                double arclength = 0;
                double linelength = 0;
                double startAngle;
                double endAngle;
                double bulge;
                double theta;
                double c;
                double r;
                double gamma;
                double phi;
                double holdangle;
                
                Vector2 p1;
                Vector2 p2;
                Vector2 center;

                foreach (Arc myarc in entities.OfType<Arc>())
                {
                    r = myarc.Radius;
                    startAngle = myarc.StartAngle;
                    endAngle = myarc.EndAngle;
                    holdangle = endAngle - startAngle;
                    arclength = (2 * Math.PI) * r * (holdangle / 360);
                    totallength = totallength + arclength;
                }   

                foreach (Line line in entities.OfType<Line>())
                {
                    vertsx = line.StartPoint.X;
                    vertsy = line.StartPoint.Y;
                    vertsz = line.StartPoint.Z;
                    vertex = line.EndPoint.X;
                    vertey = line.EndPoint.Y;
                    vertez = line.EndPoint.Z;

                    linelength = arclength = Math.Sqrt((Math.Pow((vertex - vertsx), 2) + Math.Pow((vertey - vertsy), 2) + Math.Pow((vertez - vertsz), 2)));
                    totallength = totallength + linelength;
                }

                foreach (Polyline2D poly2d in entities.OfType<Polyline2D>())
                {

                    vertextcount = poly2d.Vertexes.Count;
                    new Polyline2DVertex(new Vector2(0, 0));

                    vertx1 = poly2d.Vertexes[0].Position.X;
                    verty1 = poly2d.Vertexes[0].Position.Y;
                    vertx2 = poly2d.Vertexes[1].Position.X;
                    verty2 = poly2d.Vertexes[1].Position.Y;
                    bulge = poly2d.Vertexes[0].Bulge;

                    p1 = new Vector2(vertx1, verty1);
                    p2 = new Vector2(vertx2, verty2);

                    theta = 4 * Math.Atan(Math.Abs(bulge));
                    c = Vector2.Distance(p1, p2) / 2.0;
                    r = c / Math.Sin(theta / 2.0);

                    //i have the radius, finally again
                    gamma = (Math.PI - theta) / 2;
                    phi = Vector2.Angle(p1, p2) + Math.Sign(bulge) * gamma;
                    center = new Vector2(p1.X + r * Math.Cos(phi), p1.Y + r * Math.Sin(phi));

                    if (!MathHelper.IsZero(bulge))
                    {
                        startAngle = MathHelper.RadToDeg * Vector2.Angle(p1 - center);
                        endAngle = startAngle + MathHelper.RadToDeg * theta;
                        holdangle = endAngle - startAngle;
                        arclength = (2 * Math.PI) * r * (holdangle / 360);
                        totallength = totallength + arclength;
                    }
                    else
                    {
                        linelength = Math.Sqrt((Math.Pow((vertx1 - vertx2), 2)) + Math.Pow((verty1 - verty2), 2));
                        totallength = totallength + linelength;
                    }

                    /////////////////////////////////////////////////////////////////

                    for (int z = 1; z < vertextcount - 1; z++)
                    {
                        vertextcount = poly2d.Vertexes.Count;
                        new Polyline2DVertex(new Vector2(0, 0));

                        vertx1 = poly2d.Vertexes[z].Position.X;
                        verty1 = poly2d.Vertexes[z].Position.Y;
                        vertx2 = poly2d.Vertexes[z + 1].Position.X;
                        verty2 = poly2d.Vertexes[z + 1].Position.Y;
                        bulge = poly2d.Vertexes[z].Bulge;

                        p1 = new Vector2(vertx1, verty1);
                        p2 = new Vector2(vertx2, verty2);

                        theta = 4 * Math.Atan(Math.Abs(bulge));
                        c = Vector2.Distance(p1, p2) / 2.0;
                        r = c / Math.Sin(theta / 2.0);

                       
                        gamma = (Math.PI - theta) / 2;
                        phi = Vector2.Angle(p1, p2) + Math.Sign(bulge) * gamma;
                        center = new Vector2(p1.X + r * Math.Cos(phi), p1.Y + r * Math.Sin(phi));

                        if (!MathHelper.IsZero(bulge))
                        {
                            startAngle = MathHelper.RadToDeg * Vector2.Angle(p1 - center);
                            endAngle = startAngle + MathHelper.RadToDeg * theta;
                            holdangle = endAngle - startAngle;
                            arclength = (2 * Math.PI) * r * (holdangle / 360);
                            totallength = totallength + arclength;
                        }
                        else
                        {
                            linelength = Math.Sqrt((Math.Pow((vertx1 - vertx2), 2)) + Math.Pow((verty1 - verty2), 2));
                            totallength = totallength + linelength;
                        }
                    }
                    ///////////////////////////////////////////////////////////////// is a closed polyline?
                    if (poly2d.IsClosed)
                    {
                        vertextcount = poly2d.Vertexes.Count;
                        new Polyline2DVertex(new Vector2(0, 0));

                        vertx1 = poly2d.Vertexes[vertextcount-1].Position.X;
                        verty1 = poly2d.Vertexes[vertextcount-1].Position.Y;
                        vertx2 = poly2d.Vertexes[0].Position.X;
                        verty2 = poly2d.Vertexes[0].Position.Y;
                        bulge = poly2d.Vertexes[vertextcount-1].Bulge;

                        p1 = new Vector2(vertx1, verty1);
                        p2 = new Vector2(vertx2, verty2);

                        theta = 4 * Math.Atan(Math.Abs(bulge));
                        c = Vector2.Distance(p1, p2) / 2.0;
                        r = c / Math.Sin(theta / 2.0);

                        gamma = (Math.PI - theta) / 2;
                        phi = Vector2.Angle(p1, p2) + Math.Sign(bulge) * gamma;
                        center = new Vector2(p1.X + r * Math.Cos(phi), p1.Y + r * Math.Sin(phi));

                        if (!MathHelper.IsZero(bulge))
                        {
                            startAngle = MathHelper.RadToDeg * Vector2.Angle(p1 - center);
                            endAngle = startAngle + MathHelper.RadToDeg * theta;
                            holdangle = endAngle - startAngle;
                            arclength = (2 * Math.PI) * r * (holdangle / 360);
                            totallength = totallength + arclength;
                        }
                        else
                        {
                            linelength = Math.Sqrt((Math.Pow((vertx1 - vertx2), 2)) + Math.Pow((verty1 - verty2), 2));
                            totallength = totallength + linelength;
                        }
                    }
                    poly2d.Normal = new Vector3(0, 0, 1);
                }
                totallength = Math.Round(totallength * roundamount, 0) / roundamount;
                tempstring = totallength.ToString();
                textBox2.Text = tempstring;
                textBox2.Refresh();`

**Reply by atemmler** [reply]
Hey David,
just tried your code and it works like a charm for detection of lines in my dxf files. The other functions might also come in handy for future challenges. Thanks a lot!
Andre.

**Comment by haplokuon** [comment]
The first place to start is the TestDxfDocument project that you will find with the source code. There you will find lots of samples that I use for testing. Start executing the console application, it calls a method called "Test" that shows on screen some information of the provided DXF file "sample.dxf", it will show you how to access some of the basic data of the DXF. Some knowledge of AutoCad is advisable, since the DXF is no more than a representation of its internal data structure, and I am maintaining the same nomenclature. The AutoCad documentation is available online. 

Here you have an example:
```c#
// layer for all the lines
Layer baseLayer = new Layer("MyLayer")
{
    Color = AciColor.Red
};

// create a DXF document, in this case as AutoCad2018 version
DxfDocument doc1 = new DxfDocument(DxfVersion.AutoCad2018);

// create 300 lines
for (int i = 0; i < 300; i++)
{
    // create a line and add it to the document
    Line line = new Line(new Vector3(0, i, 0), new Vector3(5, i, 0))
    {
        Layer = baseLayer // assign the line to a layer, by default it will be layer "0"
    };
    doc1.Entities.Add(line);
}

// save the document
doc1.Save("MyFile.dxf");

// load your own DXF file or use the one just created
DxfDocument doc2 = DxfDocument.Load("MyFile.dxf");

// array for the lines end points
List<Vector3> points = new List<Vector3>();

// iterate through all objects in the active layout "doc2.Entities.ActiveLayout", by default it is the ModelSpace
// if lines are all you want you can directly iterate through "doc2.Entities.Lines" instead of "All"
foreach (EntityObject entity in doc2.Entities.All)
{
    // select lines, there are several ways to do this. This is just one of them
    if (entity is Line line)
    {
        // read and store the end points
        points.Add(line.StartPoint);
        points.Add(line.EndPoint);
    }
}
```

**Reply by atemmler** [reply]
Hello Daniel,
thanks the examples for setting the base layer, choosing the color in the base layer and creating a specific dxf version. These are super helpful for the start. 
However, the code didn't work at first since there was an error message while compiling saying 'netDxf.DxfDocument' does not contain a definition for 'Entities'. 
It took me a little while until I figured out that it should be 

doc.AddEntity(line); 
instead of 
doc.Entities.Add(Line);

Additionally, there is the same error message while compiling for the command line
foreach (EntityObject entity in doc2.Entities.All)
since netDxf.DxfDocument' does not contain a definition for 'Entities' 

Maybe you can help me to correct this line as well.

**Reply by haplokuon** [reply]
You are using an old version of the library, download the latest available code at Gibhub and compile it.

**Reply by atemmler** [reply]
Ok, thanks.

**Reply by atemmler** [reply]
I compiled the new dll and the code works now. This was a great help.

Currently, I am trying to set layers with different colors, but the mandatory mapping to ACI colors is a little bit confusings, since I try to create a continious color spectrum. I already found the AciColor.FromTrueColor(int color) color mapping function and used it as follows:

// helper funtion to create color spectrum and return int color number

	public static int Rainbow(int Count, int CountMax)
		{
		float progress = Count / CountMax;
		int ProgressColor = 0;
		int ColorNumber = 0;
		int Red = 0;
		int Green = 0;
		int Blue = 0;
		int [,] ColorPalette = new int [1300,4];
		
		for (int q = 0; q < 5; q++)
		{
			for (int i = 0;  i < 255; i++)
			{
				if (q == 0) 
				{
					Red = 255; 
					Green = i; 
					Blue = 0;
				}
				else if (q == 1) 
				{
					Red = 255 - i; 
					Green = 255; 
					Blue = 0;
				}
				else if (q == 2) 
				{
					Red = 0; 
					Green = 255; 
					Blue = i;
				}
				else if (q==3) 
				{
					Red = 0; 
					Green = 255 - i; 
					Blue = 255;
				}
				else if (q==4) 
				{
					Red = i; 
					Green = 0; 
					Blue = 255;
				}
			
				ColorPalette [q*256+i, 1] = Red;
				ColorPalette [q*256+i, 2] = Green;
				ColorPalette [q*256+i, 3] = Blue;
			}
		}
		ProgressColor = (int) (Count * 1280) / CountMax ;
		
		Red = ColorPalette[ProgressColor, 1];
		Green = ColorPalette[ProgressColor, 2];
		Blue = ColorPalette[ProgressColor, 3];
		
		ColorNumber = (255 * 255 * Red + 255 * Green + Blue);
		return ColorNumber;

		}

// main programm

override public void Main()
        {
            DxfDocument doc2 = new DxfDocument(DxfVersion.AutoCad2000);
            List<Layer> ColorLayers = new List<Layer>();

             for (int i = 0; i < iMax; i++)
            {
                ColorLayers.Add(new Layer (i.ToString()));
                ColorLayers[i].Color = AciColor.FromTrueColor(Rainbow(i, iMax));

                Line line = new Line(new Vector3(i % 2, i, 300 - i), new Vector3(i, i * i, 0))
                {
                    Layer = ColorLayers[i] 
                };

                doc2.Entities.Add(line);
            }
        }


The code works and applies different colors to each layer. However, it is not a continous color spectrum but rather a color mix. Is there a possibillity to apply true colors as layer colors rather than true colors mapped on ACI colors?

**Comment by haplokuon** [comment]
I have not thoroughly tested your code, but reading it I see a problem on how you are calculating your "ColorNumber" variable. It can be calculated as:
`int ColorNumber= BitConverter.ToInt32(new byte[] { blue, green, red, alpha}, 0);`
The alpha value does not matter since the DXF AciColor does not support transparency, it is handle through the "Transparency" property of entities and layers. The "ColorNumber" value should be the same as the one returned by the "System.Drawing.Color.ToArgb()" method.

But since, it seems, that you are trying to create a rainbow set of colors, it is much easier to use the "AciColor.FromHsl" method, just remember to re-download the code. I have added a fix that affects this method. Here there is an example:
```c#
DxfDocument doc = new DxfDocument();

const int numColors = 255;
const double saturation = 1.0;
const double lightness = 0.5;

double step = 1.0 / numColors;
for (int i = 0; i < numColors; i++)
{
    double hue = i * step;

    AciColor color = AciColor.FromHsl(hue, saturation, lightness);
    Layer layer = new Layer("Layer" + i.ToString("000"))
    {
        Color = color
    };
    doc.Layers.Add(layer);
}

doc.Save("test.dxf");
```
Another detail, The layer "0" is a default layer that must always be present in the document and it will take preference over any other layer "0" associated to an entity that you try to add to the document. All named objects, all classes that derive from TableObject, must be unique, there cannot be two with the same name. This is the case of layers, line types, styles,..., It is similar as when you copy an entity into another document, the named objects in the destination document will have preference over the ones on the original drawing, so any duplicates will have to be solved manually if necessary.

**Reply by atemmler** [reply]
Thanks, the hint towards the "System.Drawing.Color.ToArgb()" method was very helpful.  The"AciColor.FromHsl" method was not exactly what I was looking for, although it was easy to implement: In the end I used the following function to create a rainbow color spectrum (6x256 colors):

	public static Color Rainbow(int progress)
		{
		float div = (Math.Abs(progress / 255));
		int ascending = (int) ((progress % 255));
		int descending = 255 - ascending;

		switch ((int) div) //RGB
			{
			case 0:
			return Color.FromArgb(255, descending, descending, 255);
			case 1:
			return Color.FromArgb(255, 0, ascending, 255);
			case 2:
			return Color.FromArgb(255, 0, 255, descending);
			case 3:
			return Color.FromArgb(255, ascending, 255, 0);
			case 4:
			return Color.FromArgb(255, 255, descending, 0);
			case 5:
			return Color.FromArgb(255, 255, 0, ascending);
			default: // case 5:
			return Color.FromArgb(255, 0, 0, 0);
			}
		}

in the main programm I used in a i-loop:

		  if (dxf) 
		  {
		  	ColorLayers.Add(new Layer(i.ToString()));
		  	int progress = ((6*256 * i / layer));
		  	int ColorNumber = Rainbow(progress).ToArgb();
		  	ColorLayers[i].Color = AciColor.FromTrueColor(ColorNumber);
		  }

Still what is alittle bit puzzling for me:
Why does the AciColor.FromTrueColor(int24) method require an 24bit integer instead of a 32bit integer as suggested by
int ColorNumber= BitConverter.ToInt32(new byte[] { blue, green, red, alpha}, 0);?
So, the question is why "ColorLayers[i].Color = AciColor.FromTrueColor(ColorNumber);" actually works, since "System.Drawing.Color.ToArgb()" method is an int32 and AciColor.FromTrueColor() requires int24?

**Comment by haplokuon** [comment]
The AciColor is a 24 bit color since only requires the rgb channels; but the methods FromTrueColor and ToTrueColors methods use/returns 32 bit integer, and the byte for the alpha is not used whatever its value. That is how it is saved in the DXF, but you should not worry about those two methods, there is not even a reason for you to use them. If you want to manually calculate the rgb values, just initialize a new AciColor with those values, just use the constructor you prefer, passing 3 bytes or 3 doubles.
